logger/analytics/analyticslogger.go

```diff
@@ -89,22 +138,77 @@ func (al *Logger) Write(bs []byte) (int, error) {
 		return 0, err
 	}
 	bs = append(bs, '\n')
-	r := retrier.New(retrier.ExponentialBackoff(5, 100*time.Millisecond), RequestErrorClassifier{})
-	if err := r.Run(func() error {
-		_, err := al.fhAPI.PutRecord(&firehose.PutRecordInput{
-			DeliveryStreamName: aws.String(al.fhStream),
-			Record:             &firehose.Record{Data: bs},
-		})
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return 0, fmt.Errorf("PutRecords: %v", err)
+	al.mu.Lock()
+	al.batchBytes += len(bs)
+	al.batch = append(al.batch, &firehose.Record{Data: bs})
+	shouldSendBatch := len(al.batch) == al.maxBatchRecords || al.batchBytes > int(0.9*float64(al.maxBatchBytes))
+	if shouldSendBatch {
+		batch := al.batch
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+		al.wg.Add(1)
+		// be careful not to send al.batch, since we will unlock before we finish sending the batch
+		go sendBatch(&al.wg, al.errLogger, batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
 	}
+	al.mu.Unlock()
 	return len(bs), nil
 }
 
+// Close flushes all logs to Firehose.
+func (al *Logger) Close() error {
+	al.mu.Lock()
+	if len(al.batch) > 0 {
+		batch := al.batch
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+		al.wg.Add(1)
+		// be careful not to send al.batch, since we will unlock before we finish sending the batch
+		sendBatch(&al.wg, al.errLogger, batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
+	}
+	al.mu.Unlock()
+	al.wg.Wait()
+	return nil
+}
+
+func sendBatch(wg *sync.WaitGroup, errLogger logger.KayveeLogger, batch []*firehose.Record, fhAPI firehoseiface.FirehoseAPI, fhStream string, timeout time.Time) {
```

Comment:
It seems really strange to pollute the `sendBatch` api with the logger and WaitGroup. Consider removing those dependencies and wrapping `sendBatch` more effectively since these are artifacts of the `Logger` implementation.

Reply:
I did that intentionally in order to make it extra hard for people to accidentally reuse the batch slice on the logger, rather than a copy. Open to not doing that though, if you think that's overkill.